src/python/turicreate/_deps/minimal_import.py

```diff
@@ -0,0 +1,367 @@
+# -*- coding: utf-8 -*-
```

Comment:
What is the purpose of this file?

Reply:
copied from __init__.py under `_deps`. I don't know either.

Reply:
If you don't know why it's here, then let's remove it.

Reply:
Or what's going on with this file?  Are you saying there isn't a reason it's in this spot specifically?  This is a lot of code to introduce here, so we should be clear we need it.

Reply:
```
# -*- coding: utf-8 -*-
```
I think Toby concerns about this line, not the rest of the code.

Reply:
The header is 100% standard -- I think we are both wondering what the purpose of this code is and if it's necessary.  

I honestly think the full lazy-loading approach here is a bit too heavyweight for what we're hoping to do, since we're only worried about delaying the load of third party packages.   It adds more complexity than just wrapping the individual imports in a function to make it clear they are lazily loaded.

Let's just do this.  Here

```
def import_tensorflow(): 
  try: 
    import tensorflow
  except ImportError:
    # Raise helpful error message
  return tensorflow
```
Then instead of `import tensorflow`, we can just do `tensorflow = import_tensorflow()`  This would be a much more maintainable way of doing it.  

When we look at lazily import the third party modules, then we could use this machinery.

Reply:
My concerns are about the whole file, not this line.

I agree with Hoyt. The lazy loading approach seems too heavyweight. I think we should do Hoyt's suggestion instead.

Reply:
This is not doable. Since we use `pytest -m minimal`, it will run all imports first. And if we follow this path, all module depends on module level `import tensorflow` will throw. One work around is to move all `import tensorflow` to function level to do the "lazy" import. I'm not sure that if we want to go this path together with other packages.

Reply:
What you describe as the workaround sounds to me like the correct way to do lazy importing. "To avoid importing heavyweight dependencies that are only used sometimes" seems to be a primary reason to do imports within functions at all

Reply:
How do you define heavyweight? The perf overhead is very small. What people's concern is about maintenance costs. It's been tested for #3139 and also this PR. If we in the future want to roll out lazy import to all of our packages, why not we first fully test on these 3rd party dependencies and release it? Then we can have more confidence in using this method for other modules.

is this complicated? no.

This is the second time for me to tweak lazy import for minimal packages (first time #3139 ). We want to hold it back again. Let's plan it first before I do it again!

what's the deal for module-level import such as

```
import tensorflow.compat.v1 as _tf

_tf.disable_v2_behavior()
```

should I move those 2 lines to each place there `_tf` is used? Is this more maintainable than have one place using this lazy import?

Reply:
To me, this problem seems similar to some of the problems surrounding static initialization in C++ (e.g. https://isocpp.org/wiki/faq/ctors#static-init-order). Both to avoid the static initialization order fiasco, and to make initialization lazy, it's often best to defer initialization until the thing being initialized is used.

The best practice for this, in both C++ and Python, seems to be simply to perform the initialization in some function, so that the initialization only occurs when the function that needs it calls it. This is simple and easy to understand, and is self-documenting.

Building custom machinery around this to me sounds like trying to create a C++ wrapper around arbitrary types that avoids calling the constructor until it's actually used. Isn't it simpler just to do what the language supports directly?

Reply:
Anyway, I will go ahead and do the labor work.

Reply:
> To me, this problem seems similar to some of the problems surrounding static initialization in C++ (e.g. https://isocpp.org/wiki/faq/ctors#static-init-order). Both to avoid the static initialization order fiasco, and to make initialization lazy, it's often best to defer initialization until the thing being initialized is used.
> 
> The best practice for this, in both C++ and Python, seems to be simply to perform the initialization in some function, so that the initialization only occurs when the function that needs it calls it. This is simple and easy to understand, and is self-documenting.
> 
> Building custom machinery around this to me sounds like trying to create a C++ wrapper around arbitrary types that avoids calling the constructor until it's actually used. Isn't it simpler just to do what the language supports directly?

Thanks for sharing this. I don't know there's actually called Construct On First Use Idiom. I use it all the time but don't know its name. Expanding this topic, there's also destruct order issues on static objects. I've seen some in our codebase too. I manually fix with dependency injection. So I think these 2 are twin problems. 

Back to the point, the initialization is done by `import`. So the initialization dependency topology is constructed by different imports. And `import` implementation is using this `Construct On First Use Idiom` by checking the singleton in `sys.modules` (similar to static Foo* bar = new Foo(); return &*bar). Besides that, you can think `import` as a **factory function** instead of the **constructor**. There's no risk to cause `static initialization fiasco`. 

The main concern here is still the maintenance and I fully understand the concern. Thanks for your knowledge sharing, that's really helpful for me.

Reply:
For more information, `import` function uses reentrent lock to protect the read and write to the singleton map `sys.modules`.